src/lib/template-instance.ts

```diff
@@ -55,12 +56,47 @@ export class TemplateInstance {
   }
 
   _clone(): DocumentFragment {
-    // Clone the node, rather than importing it, to keep the fragment in the
-    // template's document. This leaves the fragment inert so custom elements
-    // won't upgrade and potentially modify their contents before we traverse
-    // the tree.
-    const fragment =
-        this.template.element.content.cloneNode(true) as DocumentFragment;
+    // There are a number of steps in the lifecycle of a template instance's
+    // DOM fragment:
+    //  1. Clone - create the instance fragment
+    //  2. Adopt - adopt into the main document
+    //  3. Process - find part markers and create parts
+    //  4. Upgrade - upgrade custom elements
+    //  5. Update - set node, attribute, property, etc., values
+    //  6. Connect - connect to the document. Optional and outside of this
+    //     method.
+    //
+    // We have a few constraints on the ordering of these steps:
+    //  * We need to upgrade before updating, so that property values will pass
+    //    through any property setters.
+    //  * We would like to process before upgrading so that we're sure that the
+    //    cloned fragment is inert and not disturbed by self-modifying DOM.
+    //  * We want custom elements to upgrade even in disconnected fragments.
+    //
+    // Given these constraints, with full custom elements support we would
+    // prefer the order: Clone, Process, Adopt, Upgrade, Update, Connect
+    //
+    // But Safari dooes not implement CustomElementRegistry#upgrade, so we
+    // can not implement that order and still have upgrade-before-update and
+    // upgrade disconnected fragments. So we instead sacrifice the
+    // process-before-upgrade constraint, since in Custom Elements v1 elements
+    // must not modify their light DOM in the constructor. We still have issues
+    // when co-existing with CEv0 elements like Polymer 1, and with polyfills
+    // that don't strictly adhere to the no-modification rule because shadow
+    // DOM, which may be created in the constructor, is emulated by being placed
+    // in the light DOM.
+    //
+    // The resulting order is on native is: Clone, Adopt, Upgrade, Process,
+    // Update, Connect. document.importNode() performs Clone, Adopt, and Upgrade
+    // in one step.
+    //
+    // The Custom Elements v1 polyfill supports upgrade(), so the order when
+    // polyfilled is the more ideal: Clone, Process, Adopt, Upgrade, Update,
+    // Connect.
+
+    const fragment = isCEPolyfill ?
```

Comment:
Why isn't this just switched on existence of `customElements.upgrade` (and fall back to `importNode` when no upgrade exists)?

Reply:
I considered that but that means we'd have slightly different behavior across browsers with native custom elements, rather than different between native and polyfill. Since this is also the released ordering, I thought it was safer.

Reply:
Ok so after discussion, since browsers on CE polyfill also likely need to work with ShadyDOM, keeping CE polyfill using `upgrade` offers better compatibility with elements that create polyfilled "ShadyRoots" in the constructor (which is only possible on CE polyfill because light DOM children created sync to constructor by Shady DOM would be prohibited under native CE).

This isn't an issue with native SD+CE because creating real ShadowRoots is legal in the constructor, and also hidden from the lit-html treewalk, hence we're biasing the native case to `importNode` to avoid Safari's lack of upgrade, incur fewer treewalks, and as a more conservative change from the latest released state for native browsers.